File: module/service-grpc/service_kv.go
package main

import (
	"fmt"

	"github.com/pantopic/wazero-grpc-server/sdk-go"
	"github.com/pantopic/wazero-grpc-server/sdk-go/codes"

	internal "github.com/pantopic/turbokube/module/service-grpc/internal"
)

var (
	rangeRequest = &internal.RangeRequest{}
)

func kvRange(in []byte) (err error) {
	err = rangeRequest.UnmarshalVT(in)
	if err != nil {
		grpc_server.SendErr(codes.InvalidArgument, []byte(err.Error()))
		return
	}
	out, err := grpcError(kvShard().Read(append(in, QUERY_KV_RANGE), rangeRequest.Serializable))
	if err != nil {
		fmt.Printf("Failed range request: %v\n%#v\n", err, rangeRequest)
	}
	return autoSend(out, err)
}

func kvPut(in []byte) (err error) {
	return autoSend(grpcError(kvShard().Apply(append(in, CMD_KV_PUT))))
}

func kvDeleteRange(in []byte) (err error) {
	return autoSend(grpcError(kvShard().Apply(append(in, CMD_KV_DELETE_RANGE))))
}

func kvTxn(in []byte) (err error) {
	return autoSend(grpcError(kvShard().Apply(append(in, CMD_KV_TXN))))
}

func kvCompact(in []byte) (err error) {
	return autoSend(grpcError(kvShard().Apply(append(in, CMD_KV_COMPACT))))
}

File: module-zig/service-grpc/service_kv.zig
//! Mirrors module/service-grpc/service_kv.go

const std = @import("std");
const pb = @import("pb/etcdserverpb.pb.zig");

const errors = @import("error.zig");
const module = @import("module.zig");
const types = @import("types.zig");
const util = @import("util.zig");

var scratch: [64 * 1024]u8 = undefined;
var fba = std.heap.FixedBufferAllocator.init(&scratch);

pub fn range(in: []const u8) anyerror!void {
    fba.reset();
    var reader = std.Io.Reader.fixed(in);
    const req = pb.RangeRequest.decode(&reader, fba.allocator()) catch |err| {
        try module.server.sendErr(errors.code_invalid_argument, @errorName(err));
        return;
    };
    return util.kvShard().AsyncRead(util.withSuffix(in, types.QUERY_KV_RANGE), "kvRange", req.serializable);
}

pub fn put(in: []const u8) anyerror!void {
    return util.kvShard().AsyncApply(util.withSuffix(in, types.CMD_KV_PUT), "kvPut");
}

pub fn deleteRange(in: []const u8) anyerror!void {
    return util.kvShard().AsyncApply(util.withSuffix(in, types.CMD_KV_DELETE_RANGE), "kvDeleteRange");
}

pub fn txn(in: []const u8) anyerror!void {
    return util.kvShard().AsyncApply(util.withSuffix(in, types.CMD_KV_TXN), "kvTxn");
}

pub fn compact(in: []const u8) anyerror!void {
    return util.kvShard().AsyncApply(util.withSuffix(in, types.CMD_KV_COMPACT), "kvCompact");
}
